### ISSUE TITLE
[Bug]: Additional exception handling needed in retriever.py

### ISSUE THREAD

**Issue by mphipps2** [run-llama/llama_index#7791]
### Bug Description

This line from _build_node_list_from_query_result in retriever.py:

node_ids = [
    self._index.index_struct.nodes_dict[idx] for idx in query_result.ids if idx in self._index.index_struct.nodes_dict
]

causes a crach and returns a KeyError: '-1' when the number of chunks is smaller than the length k from similarity_top_k=k. This happens because query_result.ids =  ['0', '1', '-1'] --- so that '-1' causes the crash then. This should be an easy fix, something like this would do the trick: 

****Edit: (10/03/23) fixing previous typo. The correct solution is: ****
            node_ids = [
                self._index.index_struct.nodes_dict[idx] for idx in query_result.ids if idx in self._index.index_struct.nodes_dict
            ]

or at the very least by having some sort of exception handling logic in place

### Version

0.8.30

### Steps to Reproduce

Running a simple document qa pipeline with a custom LLM class, a local embedding model and using FAISS

### Relevant Logs/Tracbacks

```shell
Traceback (most recent call last):
File "/python3.9/site-packages/llama_index/indices/vector_store/retrievers/retriever.py", line 165, in _get_nodes_with_embeddings
    return self._build_node_list_from_query_result(query_result)
File "/python3.9/site-packages/llama_index/indices/vector_store/retrievers/retriever.py", line 129, in <listcomp>
    self._index.index_struct.nodes_dict[idx] for idx in query_result.ids
KeyError: '-1'
```

**Comment by mphipps2** [comment]
Sorry, there was a typo in my original post. Inside the _build_node_list_from_query_result() function in retriever.py, the original line was:

            node_ids = [
                self._index.index_struct.nodes_dict[idx] for idx in query_result.ids
            ]

This needs changed to:

            node_ids = [
                self._index.index_struct.nodes_dict[idx] for idx in query_result.ids if idx in self._index.index_struct.nodes_dict
            ]

The solution provided for the bot doesn't work (I believe it's just the original code). So to summarize, to fix this problem this is needed:

    def _build_node_list_from_query_result(
        self, query_result: VectorStoreQueryResult
    ) -> List[NodeWithScore]:
        if query_result.nodes is None:
            # NOTE: vector store does not keep text and returns node indices.
            # Need to recover all nodes from docstore
            if query_result.ids is None:
                raise ValueError(
                    "Vector store query result should return at "
                    "least one of nodes or ids."
                )
            assert isinstance(self._index.index_struct, IndexDict)

            node_ids = [
                self._index.index_struct.nodes_dict[idx] for idx in query_result.ids if idx in self._index.index_struct.nodes_dict
            ]

            nodes = self._docstore.get_nodes(node_ids)
            query_result.nodes = nodes
        else:
            # NOTE: vector store keeps text, returns nodes.
            # Only need to recover image or index nodes from docstore
            for i in range(len(query_result.nodes)):
                source_node = query_result.nodes[i].source_node
                if (not self._vector_store.stores_text) or (
                    source_node is not None and source_node.node_type != ObjectType.TEXT
                ):
                    node_id = query_result.nodes[i].node_id
                    if self._docstore.document_exists(node_id):
                        query_result.nodes[
                            i
                        ] = self._docstore.get_node(  # type: ignore[index]
                            node_id
                        )

        log_vector_store_query_result(query_result)

        node_with_scores: List[NodeWithScore] = []
        for ind, node in enumerate(query_result.nodes):
            score: Optional[float] = None
            if query_result.similarities is not None:
                score = query_result.similarities[ind]
            node_with_scores.append(NodeWithScore(node=node, score=score))

        return node_with_scores

**Comment by bruhacsA** [comment]
I needed to apply the fix manually because it is still not updated in the PR. My retriever.py was also still the old version from last year. 

The error only occurs when no chunks are returned at all, i.e. when a search query contains a particular combination of words that somehow don't yield any results. I found that if I modified or omitted just one word from my query text, the retriever would find chunks again and this problem wouldn't surface. So it depends also on the query text.

I am wondering if it also depends on how the vector store and index was built. I am using faiss-CPU with the "BAAI/bge-small-en-v1.5" from Huggingface
